discovery_engine_core/core/src/stack/ops/common.rs

```diff
@@ -12,58 +12,32 @@
 // You should have received a copy of the GNU Affero General Public License
 // along with this program.  If not, see <https://www.gnu.org/licenses/>.
 
-use futures::{stream::FuturesUnordered, StreamExt};
 use tokio::task::JoinHandle;
-use xayn_discovery_engine_providers::Market;
 
 use crate::engine::GenericError;
 
 type ItemsResult<I> = Result<Vec<I>, GenericError>;
-type Requests<I> = FuturesUnordered<JoinHandle<ItemsResult<I>>>;
-
-async fn request_new_items<I: Send>(
-    requests_fn: impl FnOnce() -> Requests<I> + Send,
-) -> ItemsResult<I> {
-    let mut requests = requests_fn();
-    let mut items = Vec::new();
-    let mut error = None;
-
-    while let Some(handle) = requests.next().await {
-        // should we also push handle errors?
-        if let Ok(result) = handle {
-            match result {
-                Ok(batch) => items.extend(batch),
-                Err(err) => {
-                    error.replace(err);
-                }
-            }
-        }
-    }
-
-    if items.is_empty() && error.is_some() {
-        Err(error.unwrap(/* nonempty error */))
-    } else {
-        Ok(items)
-    }
-}
+type Request<I> = JoinHandle<ItemsResult<I>>;
```

Comment:
🌱 Since we only run a single request we could just return a `Future`. I just tried to replace `JoinHandle` with `Future` but it seems that it requires a bit more thinking and it is probably not worth it for the moment.

Reply:
i also gave that a try with `futures::future::Lazy` which can hold closures to be awaited. this is pretty much what we need, but since we can't name the closure type this would require quite a lot of name fiddling for the type alias. at this point i stopped and went the easy way :/